llama_index/callbacks/base.py

```diff
@@ -70,40 +74,52 @@ class CallbackManager(BaseCallbackHandler, ABC):
     Args:
         handlers (List[BaseCallbackHandler]): list of handlers to use.
 
+    Usage:
+        with callback_manager.event(CBEventType.QUERY) as event:
+            event.on_start(payload={key, val})
+            ...
+            event.on_end(payload={key, val})
+
     """
 
     def __init__(self, handlers: List[BaseCallbackHandler]):
         """Initialize the manager with a list of handlers."""
         self.handlers = handlers
         self._trace_map: Dict[str, List[str]] = defaultdict(list)
-        self._trace_event_stack: List[str] = [BASE_TRACE_EVENT]
+        self._trace_event_stack: ContextVar[List[str]] = ContextVar(
+            "trace", default=[BASE_TRACE_EVENT]
+        )
         self._trace_id_stack: List[str] = []
 
     def on_event_start(
         self,
         event_type: CBEventType,
         payload: Optional[Dict[str, Any]] = None,
-        event_id: str = "",
-        **kwargs: Any
+        event_id: Optional[str] = None,
+        **kwargs: Any,
     ) -> str:
         """Run handlers when an event starts and return id of event."""
         event_id = event_id or str(uuid.uuid4())
-        self._trace_map[self._trace_event_stack[-1]].append(event_id)
+
+        parent_id = self._trace_event_stack.get()[-1]
+        self._trace_map[parent_id].append(event_id)
         for handler in self.handlers:
             if event_type not in handler.event_starts_to_ignore:
                 handler.on_event_start(event_type, payload, event_id=event_id, **kwargs)
 
         if event_type not in LEAF_EVENTS:
-            self._trace_event_stack.append(event_id)
+            current_trace_stack = self._trace_event_stack.get().copy()

```

Comment:
very nit: could you add a quick explanation on the copy() here for the future.

Reply:
Tbh I'm not sure if it's needed, was just being cautious. Will try testing without it